# Exploring eBay Car Sales Data
**In this project, I am going to explore a dataset of used cars from eBay Kleinanzeign, a classified section of the German eBay website.**
<br>
The data dictionary provided with data is as follows:
<br>

`dateCrawled` - When this ad was first crawled. All field-values are taken from this date.
<br>
`name` - Name of the car.
<br>
`seller` - Whether the seller is private or a dealer.
<br>
`offerType` - The type of listing
<br>
`price` - The price on the ad to sell the car.
<br>
`abtest` - Whether the listing is included in an A/B test.
<br>
`vehicleType` - The vehicle Type.
<br>
`yearOfRegistration` - The year in which the car was first registered.
<br>
`gearbox` - The transmission type.
<br>
`powerPS` - The power of the car in PS.
<br>
`model` - The car model name.
<br>
`kilometer` - How many kilometers the car has driven.
<br>
`monthOfRegistration` - The month in which the car was first registered.
<br>
`fuelType` - What type of fuel the car uses.
<br>
`brand` - The brand of the car.
<br>
`notRepairedDamage` - If the car has a damage which is not yet repaired.
<br>
`dateCreated` - The date on which the eBay listing was created.
<br>
`nrOfPictures` - The number of pictures in the ad.
<br>
`postalCode` - The postal code for the location of the vehicle.
<br>
`lastSeenOnline` - When the crawler saw this ad last online.
<br>

**The aim of this project is to clean the data and analyze the included used car listings.**

In [1]:
# Import numpy and Pandas library
import numpy as np
import pandas as pd
autos = pd.read_csv("autos.csv", encoding = "Latin")


In [2]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

Now let's see the first 10 data available in the dataset

In [3]:
autos.head(10)

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5  2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6  2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7  2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8  2016-03-22 16:51:34                                         Seat_Arosa   
9  2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   
5  privat   Angebot  $7,900     test         bus                2006   
6  privat   Angebot    $300     test   limousine                1995   
7  privat   Angebot  $1,990  control   limousine                1998   
8  privat   Angebot    $250     test         NaN                2000   
9  privat   Angebot    $590  control         bus                1997   

     gearbox  powerPS    model   odometer  monthOfRegistration fuelType  \
0    manuell      158   andere  150,000km                    3      lpg   
1  automatik      286      7er  150,000km                    6   benzin   
2    manuell      102     golf   70,000km                    7   benzin   
3  automatik       71   fortwo   70,000km                    6   benzin   
4    manuell        0    focus  150,000km                    7   benzin   
5  automatik      150  voyager  150,000km                    4   diesel   
6    manuell       90     golf  150,000km                    8   benzin   
7    manuell       90     golf  150,000km                   12   diesel   
8    manuell        0    arosa  150,000km                   10      NaN   
9    manuell       90   megane  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   
5    chrysler               NaN  2016-03-21 00:00:00             0   
6  volkswagen               NaN  2016-03-20 00:00:00             0   
7  volkswagen              nein  2016-03-16 00:00:00             0   
8        seat              nein  2016-03-22 00:00:00             0   
9     renault              nein  2016-03-16 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50  
5       22962  2016-04-06 09:45:21  
6       31535  2016-03-23 02:48:59  
7       53474  2016-04-07 03:17:32  
8        7426  2016-03-26 18:18:10  
9       15749  2016-04-06 10:46:35

# Cleaning the dataset

1. The dataset contains 20 columns, most of which are strings.
<br>
2. Some columns have null values, but none have more than ~20% null values.
<br>
3. The column names use camelcase instead of Python's preferred snakecase, which means we can't just replace spaces with underscores.

<br>
Let's convert the column names from camelcase to snakecase and reword some of the column names based on the data dictionary to be more descriptive.

So first of all let's define a function to rename all columns in the dataset.

In [4]:
def clean_column(col):
    col = col.replace("yearOfRegistration","registration_year")
    col = col.replace("monthOfRegistration","registration_month")
    col = col.replace("notRepairedDamage","unrepaired_damage")
    col = col.replace("dateCreated","ad_created")
    col = col.lower()
    return col

Now let's call this clean_column function to make changes :


In [5]:
cleaned_columns = []

for data in autos.columns:
    new_column = clean_column(data)
    cleaned_columns.append(new_column)
autos.columns = cleaned_columns

Let's check the changes we have made is reflected in the dataframe or not !!

In [6]:
autos.head()

datecrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offertype   price   abtest vehicletype  registration_year  \
0  privat   Angebot  $5,000  control         bus               2004   
1  privat   Angebot  $8,500  control   limousine               1997   
2  privat   Angebot  $8,990     test   limousine               2009   
3  privat   Angebot  $4,350  control  kleinwagen               2007   
4  privat   Angebot  $1,350     test       kombi               2003   

     gearbox  powerps   model   odometer  registration_month fueltype  \
0    manuell      158  andere  150,000km                   3      lpg   
1  automatik      286     7er  150,000km                   6   benzin   
2    manuell      102    golf   70,000km                   7   benzin   
3  automatik       71  fortwo   70,000km                   6   benzin   
4    manuell        0   focus  150,000km                   7   benzin   

        brand unrepaired_damage           ad_created  nrofpictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalcode             lastseen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

So, here we can see that all column names are following the Python's standard !

Now let's do some basic data exploration to determine what other cleaning tasks need to be done. Initially we will look for: - Text columns where all or almost all values are the same. These can often be dropped as they don't have useful information for analysis.

In [7]:
autos.describe()

registration_year       powerps  registration_month  nrofpictures  \
count       50000.000000  50000.000000        50000.000000       50000.0   
mean         2005.073280    116.355920            5.723360           0.0   
std           105.712813    209.216627            3.711984           0.0   
min          1000.000000      0.000000            0.000000           0.0   
25%          1999.000000     70.000000            3.000000           0.0   
50%          2003.000000    105.000000            6.000000           0.0   
75%          2008.000000    150.000000            9.000000           0.0   
max          9999.000000  17700.000000           12.000000           0.0   

         postalcode  
count  50000.000000  
mean   50813.627300  
std    25779.747957  
min     1067.000000  
25%    30451.000000  
50%    49577.000000  
75%    71540.000000  
max    99998.000000

In the above decriptive statistics, We can see that the column "nrofpictures" has same value i.e. 0.0

And also the registration_year has max value of '9999.000000' that is not possible.

We found that the price and odometer columns are numeric values stored as text. For each column we will: 
1. remove any non-numeric characters.
2. Convert the column to a numeric dtype.
3. Use DataFrame.rename() to rename the column to odometer_km.

In [9]:
#Modifying the price column :

autos["price"] = autos["price"].str.replace("$","")
autos["price"] = autos["price"].str.replace(",","")
autos["price"] = autos["price"].astype(int)

In [10]:
#Modifying the odometer column :

autos["odometer"] = autos["odometer"].str.replace(",","").str.replace("km","").astype(int)
autos.rename({"odometer": "odometer_km"}, axis=1, inplace=True)

Now we can verify the changes !

In [11]:
autos.head(20)

datecrawled                                               name  \
0   2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1   2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2   2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3   2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4   2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5   2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6   2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7   2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8   2016-03-22 16:51:34                                         Seat_Arosa   
9   2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10  2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11  2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12  2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13  2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14  2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15  2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16  2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17  2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18  2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19  2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   

    seller offertype  price   abtest vehicletype  registration_year  \
0   privat   Angebot   5000  control         bus               2004   
1   privat   Angebot   8500  control   limousine               1997   
2   privat   Angebot   8990     test   limousine               2009   
3   privat   Angebot   4350  control  kleinwagen               2007   
4   privat   Angebot   1350     test       kombi               2003   
5   privat   Angebot   7900     test         bus               2006   
6   privat   Angebot    300     test   limousine               1995   
7   privat   Angebot   1990  control   limousine               1998   
8   privat   Angebot    250     test         NaN               2000   
9   privat   Angebot    590  control         bus               1997   
10  privat   Angebot    999     test         NaN               2017   
11  privat   Angebot    350  control         NaN               2000   
12  privat   Angebot   5299  control  kleinwagen               2010   
13  privat   Angebot   1350  control   limousine               1999   
14  privat   Angebot   3999     test  kleinwagen               2007   
15  privat   Angebot  18900     test       coupe               1982   
16  privat   Angebot    350     test       kombi               1999   
17  privat   Angebot   5500     test       coupe               1990   
18  privat   Angebot    300  control         bus               1995   
19  privat   Angebot   4150  control         suv               2004   

      gearbox  powerps     model  odometer_km  registration_month fueltype  \
0     manuell      158    andere       150000                   3      lpg   
1   automatik      286       7er       150000                   6   benzin   
2     manuell      102      golf        70000                   7   benzin   
3   automatik       71    fortwo        70000                   6   benzin   
4     manuell        0     focus       150000                   7   benzin   
5   automatik      150   voyager       150000                   4   diesel   
6     manuell       90      golf       150000                   8   benzin   
7     manuell       90      golf       150000                  12   diesel   
8     manuell        0     arosa       150000                  10      NaN   
9     manuell       90    megane       150000                   7   benzin   
10    manuell       90 

Here We go !! We have updated the price and odometer_km columns as per requirement !!

Let's continue exploring the data, specifically looking for data that doesn't look right. We'll start by analyzing the odometer_km and price columns. Here's the steps we'll take:
<br>
Analyze the columns using minimum and maximum values and look for any values that look unrealistically high or low (outliers) that we might want to remove.
<br>
We'll use:
<br>
1. Series.unique().shape to see how many unique values
2. Series.describe() to view min/max/median/mean etc
3. Series.value_counts(), with some variations:
4. chained to .head() if there are lots of values.Because Series.value_counts() returns a series, we can use Series.sort_index() with ascending= True or False to view the highest and lowest values with their counts.

In [16]:
#Let's analyse the price and odometer_km !!

price = autos["price"]
odometer = autos["odometer_km"]

print("PRICE COLUMN DETAILS :")
price.value_counts()
#print("Odometr COLUMN DETAILS :")
#odometer.value_counts().head(10)

PRICE COLUMN DETAILS :


0         1421
500        781
1500       734
2500       643
1000       639
1200       639
600        531
800        498
3500       498
2000       460
999        434
750        433
900        420
650        419
850        410
700        395
4500       394
300        384
2200       382
950        379
1100       376
1300       371
3000       365
550        356
1800       355
5500       340
1250       335
350        335
1600       327
1999       322
          ... 
46200        1
29600        1
13480        1
21700        1
7373         1
3279         1
4286         1
188          1
17830        1
9130         1
910          1
238          1
2671         1
69900        1
151990       1
2479         1
4510         1
86500        1
47499        1
16998        1
27299        1
41850        1
4780         1
686          1
6495         1
20790        1
8970         1
846          1
2895         1
33980        1
Name: price, Length: 2357, dtype: int64

Here we can see that the price of 1421 car is ZERO that's not possible. So we have to remove the outlier here!

In [22]:
#To find max & min value in price column

print("MAX in price column :")
print(price.max())
print("\nMIN in price column :")
print(price.min())

MAX in price column :
99999999

MIN in price column :
0


So now we have remove the outlier here ! And we can assume that the price of car should not exceeeds the value of 100000 !

In [28]:
autos = autos[autos["price"].between(1000,100000)]

In [29]:
autos.describe()

price  registration_year       powerps    odometer_km  \
count  38587.000000       38587.000000  38587.000000   38587.000000   
mean    7089.045482        2005.680125    128.663177  122859.771426   
std     7893.814349          86.727780    215.791791   40712.638070   
min     1000.000000        1000.000000      0.000000    5000.000000   
25%     2200.000000        2001.000000     80.000000  100000.000000   
50%     4300.000000        2005.000000    116.000000  150000.000000   
75%     8950.000000        2009.000000    160.000000  150000.000000   
max    99900.000000        9999.000000  17700.000000  150000.000000   

       registration_month  nrofpictures    postalcode  
count        38587.000000       38587.0  38587.000000  
mean             5.999559           0.0  51682.914298  
std              3.591692           0.0  25689.243673  
min              0.000000           0.0   1067.000000  
25%              3.000000           0.0  31199.000000  
50%              6.000000           0.0  50829.000000  
75%              9.000000           0.0  72528.000000  
max             12.000000           0.0  99998.000000

Here in above description we can see that the minimum value of price is 1000 and the max is 99900. And in the odometer_km we don't jave to make more changes. So now we can move ahead.

# Modifying Date

Let's now move on to the date columns and understand the date range the data covers.
<br>

There are 5 columns that should represent date values. Some of these columns were created by the crawler, some came from the website itself. We can differentiate by referring to the data dictionary:
<br>
 `date_crawled`: added by the crawler
<br> 
 `last_seen`: added by the crawler
<br>
 `ad_created`: from the website
<br>
 `registration_month`: from the website
<br>
 `registration_year`: from the website
<br>

Right now, the date_crawled, last_seen, and ad_created columns are all identified as string values by pandas. 

<br>
Because these three columns are represented as strings, we need to convert the data into a numerical representation so we can understand it quantitatively. The other two columns are represented as numeric values, so we can use methods like Series.describe() to understand the distribution without any extra data processing.

Let's first understand how the values in the three string columns are formatted. These columns all represent full timestamp values, like so:

In [30]:
autos[['datecrawled','ad_created','lastseen']][0:5]

datecrawled           ad_created             lastseen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

We noticed that the first 10 characters represent the day (e.g. 2016-03-12). So now we can extract just the date values, use Series.value_counts() to generate a distribution, and then sort by the index.

In [31]:
#modifying the datecrawled, ad_created, lastseen column

autos["datecrawled"] = autos["datecrawled"].str[:10]
autos["ad_created"] = autos["ad_created"].str[:10]
autos["lastseen"] = autos["lastseen"].str[:10]



In [32]:
#Now we can check the updated values in the above columns

autos[['datecrawled','ad_created','lastseen']][0:5]

datecrawled  ad_created    lastseen
0  2016-03-26  2016-03-26  2016-04-06
1  2016-04-04  2016-04-04  2016-04-06
2  2016-03-26  2016-03-26  2016-04-06
3  2016-03-12  2016-03-12  2016-03-15
4  2016-04-01  2016-04-01  2016-04-01

In [33]:
#lets explore the "datecrawled" column :
autos["datecrawled"].value_counts(normalize=True, dropna=False)

2016-04-03    0.039132
2016-03-20    0.038173
2016-03-12    0.037396
2016-03-21    0.037266
2016-04-04    0.036904
2016-03-14    0.036670
2016-04-02    0.036333
2016-03-28    0.035349
2016-03-19    0.035167
2016-03-07    0.035167
2016-04-01    0.034571
2016-03-29    0.034001
2016-03-15    0.033612
2016-03-10    0.033353
2016-03-26    0.033094
2016-03-30    0.033016
2016-03-11    0.032835
2016-03-08    0.032602
2016-03-09    0.032472
2016-03-22    0.032420
2016-03-23    0.032213
2016-03-31    0.031410
2016-03-27    0.031384
2016-03-25    0.030528
2016-03-17    0.030477
2016-03-16    0.029077
2016-03-24    0.029025
2016-03-05    0.025527
2016-03-13    0.016016
2016-03-06    0.013891
2016-04-05    0.013346
2016-03-18    0.012802
2016-04-06    0.003265
2016-04-07    0.001503
Name: datecrawled, dtype: float64

In [34]:
autos["ad_created"].value_counts(normalize=True, dropna=False)

2016-04-03    0.039443
2016-03-20    0.038277
2016-03-21    0.037552
2016-04-04    0.037318
2016-03-12    0.037137
2016-04-02    0.035997
2016-03-28    0.035375
2016-03-14    0.035012
2016-04-01    0.034416
2016-03-19    0.034105
2016-03-29    0.033975
2016-03-07    0.033846
2016-03-15    0.033431
2016-03-26    0.033198
2016-03-11    0.033094
2016-03-10    0.033016
2016-03-30    0.032887
2016-03-09    0.032679
2016-03-08    0.032524
2016-03-22    0.032213
2016-03-23    0.031954
2016-03-31    0.031565
2016-03-27    0.031254
2016-03-25    0.030684
2016-03-17    0.030140
2016-03-16    0.029647
2016-03-24    0.029025
2016-03-05    0.023013
2016-03-13    0.017674
2016-03-06    0.015161
                ...   
2016-02-25    0.000078
2016-02-02    0.000052
2016-02-24    0.000052
2016-02-18    0.000052
2016-02-12    0.000052
2016-01-10    0.000052
2016-02-05    0.000052
2016-02-14    0.000052
2016-02-21    0.000052
2016-02-09    0.000026
2015-09-09    0.000026
2016-02-01    0.000026
2016-02-26 

In [36]:
autos["lastseen"].value_counts(normalize=True, dropna=False)

2016-04-06    0.234742
2016-04-07    0.139969
2016-04-05    0.130977
2016-03-17    0.026408
2016-04-02    0.024905
2016-04-03    0.024438
2016-03-30    0.023479
2016-04-04    0.023402
2016-04-01    0.023220
2016-03-31    0.022754
2016-03-12    0.022210
2016-03-22    0.020810
2016-03-29    0.020784
2016-03-20    0.019799
2016-03-21    0.019670
2016-03-28    0.019411
2016-03-24    0.018504
2016-03-23    0.017908
2016-03-25    0.017778
2016-03-26    0.016016
2016-03-16    0.015446
2016-03-15    0.015005
2016-03-19    0.014616
2016-03-27    0.014072
2016-03-14    0.011999
2016-03-11    0.011740
2016-03-10    0.009822
2016-03-09    0.008889
2016-03-13    0.008397
2016-03-18    0.007360
2016-03-08    0.006246
2016-03-07    0.004561
2016-03-06    0.003576
2016-03-05    0.001088
Name: lastseen, dtype: float64

In [37]:
autos["datecrawled"].sort_index()

0        2016-03-26
1        2016-04-04
2        2016-03-26
3        2016-03-12
4        2016-04-01
5        2016-03-21
7        2016-03-16
12       2016-03-31
13       2016-03-23
14       2016-03-23
15       2016-04-01
17       2016-03-29
19       2016-03-17
20       2016-03-05
21       2016-03-06
22       2016-03-28
23       2016-03-10
24       2016-04-03
28       2016-03-19
29       2016-04-02
31       2016-03-14
33       2016-03-15
34       2016-04-05
35       2016-03-29
36       2016-03-16
37       2016-03-23
38       2016-03-21
39       2016-03-11
40       2016-03-07
41       2016-03-10
            ...    
49962    2016-03-14
49963    2016-03-26
49964    2016-03-10
49965    2016-03-11
49966    2016-04-02
49967    2016-03-12
49968    2016-04-01
49969    2016-03-17
49970    2016-03-21
49972    2016-03-26
49973    2016-03-27
49975    2016-03-27
49976    2016-03-19
49977    2016-03-31
49979    2016-03-20
49981    2016-03-15
49982    2016-03-29
49985    2016-04-02
49986    2016-04-04


In [38]:
autos["registration_year"].describe()

count    38587.000000
mean      2005.680125
std         86.727780
min       1000.000000
25%       2001.000000
50%       2005.000000
75%       2009.000000
max       9999.000000
Name: registration_year, dtype: float64

One thing that stands out, that the registration_year column contains some odd values:
<br>

The minimum value is 1000, before cars were invented
<br>
The maximum value is 9999, many years into the future

<br>
Because a car can't be first registered after the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate. Determining the earliest valid year is more difficult. Realistically, it could be somewhere in the first few decades of the 1900s.
<br>

Let's count the number of listings with cars that fall outside the 1900 - 2016 interval and see if it's safe to remove those rows entirely, or if we need more custom logic.

In [39]:
autos = autos[autos["registration_year"].between(1900,2016)]

Now calculate the distribution of the remaining values in "registration_year" column : 

In [40]:
autos["registration_year"].value_counts(normalize=True)

2005    0.074932
2006    0.071327
2004    0.070170
2003    0.066645
2007    0.060753
2008    0.059246
2002    0.057443
2009    0.055856
2001    0.055533
2000    0.053811
1999    0.046385
2011    0.043479
2010    0.042511
2012    0.035058
1998    0.034439
2013    0.021282
1997    0.021148
2014    0.017408
2016    0.017354
1996    0.014583
1995    0.011892
2015    0.009605
1994    0.007291
1993    0.005865
1992    0.005839
1991    0.005569
1990    0.005085
1989    0.003309
1988    0.002825
1985    0.002126
          ...   
1976    0.000565
1977    0.000511
1975    0.000484
1969    0.000484
1965    0.000457
1960    0.000457
1964    0.000242
1963    0.000215
1959    0.000161
1961    0.000161
1956    0.000108
1962    0.000108
1937    0.000108
1958    0.000081
1954    0.000054
1955    0.000054
1941    0.000054
1957    0.000054
1934    0.000054
1953    0.000027
1951    0.000027
1943    0.000027
1927    0.000027
1929    0.000027
1931    0.000027
1950    0.000027
1948    0.000027
1938    0.0000

Here we can observe that the year 2005 has the maximum number of car registrations and the year 1952 has lowest number of car registration.

When working with data on cars, it's natural to explore variations across different car brands. We can use aggregation to understand the brand column.

# Exploring Price by Brand

In [41]:
autos["brand"].value_counts(normalize=True)

volkswagen        0.211021
bmw               0.125407
mercedes_benz     0.111631
audi              0.097667
opel              0.089165
ford              0.058735
renault           0.037318
peugeot           0.027928
fiat              0.021094
skoda             0.019076
seat              0.017300
smart             0.016628
toyota            0.014637
mazda             0.014260
citroen           0.013910
nissan            0.013641
mini              0.010897
hyundai           0.010762
sonstige_autos    0.010332
volvo             0.008986
kia               0.007695
honda             0.007345
mitsubishi        0.006888
porsche           0.006780
chevrolet         0.006619
alfa_romeo        0.006242
suzuki            0.005731
dacia             0.003282
chrysler          0.003175
jeep              0.002771
land_rover        0.002637
jaguar            0.001856
subaru            0.001722
daihatsu          0.001695
saab              0.001372
daewoo            0.000915
trabant           0.000861
r

Here we can see that Volkswagen is by far the most popular brand, with approximately double the cars for sale of the next two brands combined.

There are lots of brands that don't have a significant percentage of listings, so we will limit our analysis to brands representing more than 5% of total listings.

In [44]:
brand_counts = autos["brand"].value_counts(normalize=True)
common_brands = brand_counts[brand_counts > .05].index
print(common_brands)


Index(['volkswagen', 'bmw', 'mercedes_benz', 'audi', 'opel', 'ford'], dtype='object')


In [45]:
brand_mean_prices = {}

for brand in common_brands:
    brand_only = autos[autos["brand"] == brand]
    mean_price = brand_only["price"].mean()
    brand_mean_prices[brand] = int(mean_price)
    

In [47]:
brand_mean_prices

{'volkswagen': 6645,
 'bmw': 8975,
 'mercedes_benz': 9196,
 'audi': 10276,
 'opel': 4219,
 'ford': 5274}

Out of the top 5 brands, there is distinct price gap:
    - Audi,BMW and Mercedes Benz are more expensive
    - Ford and Opel are less expensive
    - Volkswagen is in between - this may explain its popularity, it may be a 'best of 'both worlds' option.

# Exploring Mileage

In [49]:
bmp_series = pd.Series(brand_mean_prices)
pd.DataFrame(bmp_series, columns=["mean_price"])

mean_price
volkswagen           6645
bmw                  8975
mercedes_benz        9196
audi                10276
opel                 4219
ford                 5274

In [51]:
brand_mean_mileage = {}

for brand in common_brands:
    brand_only = autos[autos["brand"] == brand]
    mean_mileage = brand_only["odometer_km"].mean()
    brand_mean_mileage[brand] = int(mean_mileage)

mean_mileage = pd.Series(brand_mean_mileage).sort_values(ascending=False)
mean_prices = pd.Series(brand_mean_prices).sort_values(ascending=False)


In [52]:
brand_info = pd.DataFrame(mean_mileage,columns=["mean_mileage"])
brand_info

mean_mileage
bmw                  132068
mercedes_benz        130130
audi                 127524
volkswagen           125765
opel                 123952
ford                 119651

In [53]:
brand_info["mean_price"] = mean_prices
brand_info

mean_mileage  mean_price
bmw                  132068        8975
mercedes_benz        130130        9196
audi                 127524       10276
volkswagen           125765        6645
opel                 123952        4219
ford                 119651        5274



Here we can observe that the range of car mileages doesn't vary as much as the prices do by brand. And there is a slight trend to the more expensive vehicles having higher mileage, with the less expensive vehicles having lower mileage.